chain/rust/src/fees.rs

```diff
@@ -0,0 +1,639 @@
+use cml_core::{ArithmeticError, serialization::Serialize};
+use fraction::{Fraction, ToPrimitive};
+use crate::Coin;
+use crate::plutus::ExUnitPrices;
+use crate::plutus::utils::compute_total_ex_units;
+use crate::transaction::Transaction;
+
+/// Careful: although the linear fee is the same for Byron & Shelley
+/// The value of the parameters and how fees are computed is not the same
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub struct LinearFee {
+    pub constant: Coin,
+    pub coefficient: Coin,
+}
+
+impl LinearFee {
+    pub fn new(coefficient: Coin, constant: Coin) -> Self {
+        Self {
+            constant,
+            coefficient,
+        }
+    }
+}
+
+pub fn min_script_fee(tx: &Transaction, ex_unit_prices: &ExUnitPrices) -> Result<Coin, ArithmeticError> {
+    if let Some(redeemers) = &tx.witness_set.redeemers {
+        let total_ex_units = compute_total_ex_units(&redeemers)?;
+        let script_fee = (
+            (
+                Fraction::new(total_ex_units.mem, 1u64)
+                * Fraction::new(
+                    ex_unit_prices.mem_price.numerator,
+                    ex_unit_prices.mem_price.denominator
+                )
+            )
+            +
+            (
+                Fraction::new(total_ex_units.steps, 1u64)
+                * Fraction::new(
+                    ex_unit_prices.step_price.numerator,
+                    ex_unit_prices.step_price.denominator,
+                )
+            )
+        ).ceil().to_u64().unwrap();
+        Ok(script_fee)
+    } else {
+        Ok(0)
+    }
+}
+
+pub fn min_no_script_fee(
+    tx: &Transaction,
+    linear_fee: &LinearFee,
+) -> Result<Coin, ArithmeticError> {
+    (tx.to_cbor_bytes().len() as u64)
+        .checked_mul(linear_fee.coefficient)
+        .and_then(|x| x.checked_add(linear_fee.constant))
+        .ok_or(ArithmeticError::IntegerOverflow)
+}
+
+pub fn min_fee(
+    tx: &Transaction,
+    linear_fee: &LinearFee,
+    ex_unit_prices: &ExUnitPrices
+) -> Result<Coin, ArithmeticError> {
+    // TODO: the fee should be 0 if all inputs are genesis redeem addresses
+    min_no_script_fee(tx, linear_fee)?
+        .checked_add(min_script_fee(tx, ex_unit_prices)?)
+        .ok_or(ArithmeticError::IntegerOverflow)
+}
+
+// Note: all the tests below are disabled
+// This is because these tests are meant to run on Alonzo-era tx formats to give the right result
+// But currently CML doesn't allow you to encode txs in old formats explicitly
+// which means these tests return the wrong value until we add this feature
```

Comment:
do we plan to have this feature?

Reply:
They were commented out in the old CML as well, I guess from when we updated to babbage. I got rid of the legacy fee calculation/etc when refactoring now. We had talked about this previously and IIRC decided that we want to support parsing of every era, but for creation/builders it makes more sense to only support the current era. I think I will just remove these commented out tests then.